plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -285,60 +529,334 @@ public boolean deleteNuageVspDevice(DeleteNuageVspDeviceCmd cmd) {
-                    if (thread.isDaemon())
-                        thread.setDaemon(false);
-                    if (thread.getPriority() != Thread.NORM_PRIORITY)
-                        thread.setPriority(Thread.NORM_PRIORITY);
-                    return thread;
+    private void auditHost(HostVO host) {
+        _hostDao.loadDetails(host);
+
+        boolean validateDomains = true;
+        List<NuageVspDeviceVO> nuageVspDevices = _nuageVspDao.listByHost(host.getId());
+        if (!CollectionUtils.isEmpty(nuageVspDevices)) {
+            for (NuageVspDeviceVO nuageVspDevice : nuageVspDevices) {
+                ConfigurationVO cmsIdConfig = _configDao.findByName("nuagevsp.cms.id");
+                String nuageVspCmsId = findNuageVspCmsIdForDevice(nuageVspDevice.getId(), cmsIdConfig);
+                SyncNuageVspCmsIdCommand syncCmd = new SyncNuageVspCmsIdCommand(SyncType.AUDIT, nuageVspCmsId);
+                SyncNuageVspCmsIdAnswer answer = (SyncNuageVspCmsIdAnswer) _agentMgr.easySend(nuageVspDevice.getHostId(), syncCmd);
+
+                if (answer != null && !answer.getSuccess()) {
+                    s_logger.error("Nuage VSP Device with ID " + nuageVspDevice.getId() + " is configured with an unknown CMS ID!");
+                    validateDomains = false;
+                } else if (answer != null && answer.getSyncType() == SyncType.REGISTER) {
+                    registerNewNuageVspDevice(cmsIdConfig, nuageVspDevice.getId() + ":" + answer.getNuageVspCmsId());
                 }
-            };
-            scheduler = Executors.newScheduledThreadPool(numOfSyncThreads, threadFactory);
-            scheduler.scheduleWithFixedDelay(new NuageVspSyncTask("FLOATING_IP"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes,
-                    TimeUnit.MILLISECONDS);
-            scheduler.scheduleWithFixedDelay(new NuageVspSyncTask("ENTERPRISE_NTWK_MACRO"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes,
-                    TimeUnit.MILLISECONDS);
-            scheduler
-                    .scheduleWithFixedDelay(new NuageVspSyncTask("ENTERPRISE"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes, TimeUnit.MILLISECONDS);
-        } else {
-            s_logger.warn("NuageVsp configuration for syncWorkers=" + numOfSyncThreads + " syncInterval=" + syncUpIntervalInMinutes
-                    + " could not be read properly. So, check if the properties are configured properly in global properties");
+            }
+        }
+
+        if (validateDomains) {
+            auditDomainsOnVsp(host, true, false);
         }
     }
 
-    public class NuageVspSyncTask implements Runnable {
+    private boolean auditDomainsOnVsp(HostVO host, boolean add, boolean remove) {
+        List<NuageVspDeviceVO> nuageVspDevices = _nuageVspDao.listByHost(host.getId());
+        if (!CollectionUtils.isEmpty(nuageVspDevices)) {
```

Comment:
To reduce method complexity, short-circuit the method if the ``nuageVspDevices`` list is empty by returning ``true`` to remove the unnecessary nesting.